src/qibolab/instruments/emulator/pulse_simulator.py

```diff
@@ -0,0 +1,711 @@
+    access to hardware.
+    """
+
+    PortType = DummyPort  # Placeholder. Copied over from dummy, may not be required.
+
+    def __init__(
+        self,
+        simulation_config: dict,
+        model_config: dict,
+        sim_opts: Optional[None] = None,
+    ):
+        """Set emulator configuration.
+
+        Args:
+            simulation_config (dict): Simulation configuration dictionary.
+            model_config (dict): Model configuration dictionary.
+            sim_opts (optional): Simulation backend specific object specifying simulation options.
+        """
+        self.simulation_config = simulation_config
+        self.model_config = model_config
+        self.sim_opts = sim_opts
+
+        self.all_simulation_backends = {"Qutip": QutipSimulator}
+
+        self.available_sweep_parameters = {
+            Parameter.amplitude,
+            Parameter.duration,
+            Parameter.frequency,
+            Parameter.relative_phase,
+            Parameter.start,
+        }
+
+        self.update()
+
+    @property
+    def sampling_rate(self):
+        return self.model_config["sampling_rate"]
+
+    def update(self):
+        """Updates the pulse simulator by loading all parameters from
+        `self.model_config` and `self.simulation_config`."""
+        self.simulation_backend_name = self.simulation_config["simulation_backend_name"]
+        self.device_name = self.model_config["device_name"]
+        self.model_name = self.model_config["model_name"]
+        self.emulator_name = f"{self.device_name} emulator running {self.model_name} on {self.simulation_backend_name} backend"
+        self.simulation_backend = self.all_simulation_backends[
+            self.simulation_backend_name
+        ](self.model_config, self.sim_opts)
+
+        self.platform2simulator_channels = self.model_config[
+            "platform2simulator_channels"
+        ]
+
+        self.sim_sampling_boost = self.simulation_config["sim_sampling_boost"]
+        self.instant_measurement = self.simulation_config["instant_measurement"]
+        self.runcard_duration_in_dt_units = self.model_config[
+            "runcard_duration_in_dt_units"
+        ]
+        # self.readout_error = self.model_config["readout_error"]
+        self.readout_error = {}
+        for k, v in self.model_config["readout_error"].items():
+            self.readout_error.update({int(k): v})
+        self.exec_params = ExecutionParameters
+        self.default_nshots = self.simulation_config["default_nshots"]
+        self.exec_params.nshots = self.default_nshots
+        self.exec_params.acquisition_type = AcquisitionType.DISCRIMINATION
+        self.simulate_dissipation = self.simulation_config["simulate_dissipation"]
+
+        self.pulse_sequence_history = []
+        self.channel_waveforms_history = []
+
+    def update_sim_opts(self, updated_sim_opts):
+        self.sim_opts = updated_sim_opts
+        self.simulation_backend.update_sim_opts(updated_sim_opts)
```

Comment:
Please, try to avoid as much as possible to use the internal state.

Having a large object with a lot of data attached, and then using its methods throughout the whole execution, is very close to just work with global variables.

In the end, there will be only one instance of `PulseSimulator` per execution (at least in most scenarios), and according to these `updates*` methods it will trace almost all the information.
Thus, any `PulseSimulator` method will be able to access almost everything, and affect the subsequent execution of any other method.
This is not modular and very coupled, and it will make it harder to understand the flow and reproduce the errors (since, potentially, you will need to reproduce the exact state, whatever it is).

Reply:
Sorry I do not understand this comment. What are you referring to as the internal state?
This `update` method is essentially just a reset and initialize initialization method. Do you mean that if there is a need to, simply initialize a different instance of `PulseSimulator` instead of resetting a current one?

Reply:
The internal state of an object is the content of all its attributes.
Quick translation: avoid assigning to `self` attributes as much as possible.

You're not using everything at once, and some information is only needed for specific methods, to which it could be passed as an argument. Some information is even not needed at all (take the `self.emulator_name` that could be generated on the fly by other attributes, but even all the other `self.*_name`, that most likely are there only for logging).
In general, it's good practice to keep the amount of data stored minimal and isolated, and communicate data through function parameters and return values (this goes in the direction of functional programming, but even object-oriented benefits by limited use of globals and global-locals).

Reply:
Removed some attributes in 7f4b9230dd3d7291758bb3681abb75570ec34bf6. I don't think I can do with any less at the moment.